### ISSUE TITLE
Installing CLIMSOFT

### ISSUE THREAD

**Issue by rdstern** [opencdms/Climsoft#229]
[Installation of CLIMSOFT v4_Cathy Garlick.docx](https://github.com/opencdms/Climsoft/files/1244588/Installation.of.CLIMSOFT.v4_Cathy.Garlick.docx)

The latest version of CLIMSOFT was supplied and the Word report (attached) describes Cathy's problems with the installation.  She wrote:
"See attached the results of my attempts at installing the software!!  If I can’t get beyond the installation then it doesn’t bode well for the rest of it!"

**Comment by rdstern** [comment]
Marcellin replied - see below - and also attached an installation guide:
[Readme_Climsoft_V4.txt](https://github.com/opencdms/Climsoft/files/1244603/Readme_Climsoft_V4.txt)

He writes:
"The installation is not straight forward but there is a readme file attached to this email which explain steps by steps how to install climsoftV4. From the attachment sent by Cathy,I realised that:
1. She did not run the database script provided in the dbase folder  and hence,the system could not communicate with the database;I am not sure if after installing the database client she created a new database instance and define the port number.
2. She did not set up user permission on climsoft folder (climsoftV4) after installation;If this is not set after installation process,the resource file can not be reached and hence the failure to login into the system (note that this is done once). All the steps are defined in the draft climsoft user's manual documentation.
3. To run the set up(executable),you don't need to be the administrator but you need to right click the setup file and select run "as Administrator". 
More discussion will be done to make sure the installation process of climsoft is easy and straight forward."

**Comment by rdstern** [comment]
I am not sure whether the instructions were supplied originally.  However installing software (at least by experienced users) is usually done now just by getting on - and reading instructions only if desperate. So I am happy with Marcellin's last comment above that the installation process will be made easy.

My question is what should we do now as the next step.  I would like to suggest another group tries the installation - and the team in Maseno is an obvious one, partly because they already have data from stations in Western Kenya in CLIMSOFT 3.2.  I assume they should wait a few days for your discussions.  Ideally they would try the installation next week (beginning 28th August) when David Stern will also be there?

**Comment by mhabimana** [comment]
Thanks @rdstern . Great if the group at Maseno will try as well the software and provide us with feedback to improve on this.

**Comment by isedwards** [comment]
@smachua, @mhabimana  - do either of you know why elevated administrator permissions are needed during the setup?

**Comment by smachua** [comment]
While installing and configuring Climsoft V4, use of the provided guidelines (manual or readme file) is necessary. The installation wizard will take you through Climsoft user interface installation and MariaDB (if the option is chosen). The rest that includes running sql scripts to create climsoft database and setting of permissions in the Windows are accomplished by following the guidelines. This requirement is common with CDMSs since their front ends and back ends are totally different systems.

**Comment by mhabimana** [comment]
@isedwards - I think elevated administrator permissionis needed during the installation because it is set  for climsoft that the installation can not be done by any users on the PC without the administrator privilege; to install properly you need to be the administrator or member of the administrators and not a simple user. @smachua can elaborate more on this.

**Comment by isedwards** [comment]
The CLIMSOFT application executes SQL commands all the time.

The easy way to improve the installation is simply to get CLIMSOFT to execute the initial setup script the first time it runs if the database has not yet been set up.  This would remove the manual step and make sure that everything is set up correctly without requiring the user to take manual steps.  Why not? What is the advantage of making a user do it manually during each setup?

It's essential that no part of Climsoft needs to run as an administrator.  If the user received an infected version of the installer they could end up installing all manner of malware, viruses, keyloggers and who knows what onto their machine that had full access to the entire machine.  As a guideline - software should be run as administrator only if it is physically altering the Windows operating system in a way that requires being an administrator, we definitely don't want to be doing this - and we don't want our users to think we are either.

**Comment by isedwards** [comment]
Didn't we agree in https://github.com/opencdms/Climsoft/issues/193 to **"Revert to MySQL rather than using MariaDB"**?  Can we update the installer to reflect this, or do we need to reopen https://github.com/opencdms/Climsoft/issues/193 and discuss this more?

**Comment by mhabimana** [comment]
Thanks @isedwards, Are you happy to create an installer using this idea for us to compare with the current?I think this will be the best way to move forward.

**Comment by rdstern** [comment]
This seems to be a useful discussion.  I assume we should wait for a new installer before asking the Maseno team to try the software?

**Comment by isedwards** [comment]
I'm happy to create an alternative installer if we all agree on the way forward.  This would ensure that a wider group of people could successfully test the current version without running into problems at the installation stage.

The steps would be:
1. Fix the minor issues in https://github.com/opencdms/Climsoft/pull/228 and merge Samuel's latest work into master
2. Switch to using [Advanced Installer](http://www.advancedinstaller.com/free-license.html) as discussed at length in early 2016. Since Climsoft is an open source project we have a free licence (see my original email on 16/05/2016). There are several benefits, but the main ones are that it handles the complexity of existing .NET dependencies better than Inno Setup and can create msi installers (see below)
3. Remove all third-party software from our install process, this simplifies the setup, makes it faster, removes difficulties (like a hidden dialog that no one is aware of popping up in the background during WRPlot install) and also hopefully may mean that we can create an msi installer instead of an exe installer which is [more secure and trustable](https://www.symantec.com/connect/articles/understanding-difference-between-exe-and-msi).:
    - MySQL/MariaDB will not be installed, but CLIMSOFT will help with this process on first-run if the user needs to install the database
    - WRPlot will not be installed - but it can still be an option within CLIMSOFT for users that want it.   If it is not currently installed then CLIMSOFT will give instructions
    - Advanced Installer will take responsibility for ensuring the .NET dependency is met without us adding it as an option for the user to select/unselect, guaranteeing that the correct version is always present
4. Revert to recommending MySQL rather than using MariaDB as per @Steve-Palmer's request in https://github.com/opencdms/Climsoft/issues/193
5. Remove the need for anything to be done using the window's Administrator account, including implementing an alternative to the config.inf file that also requires elevated privileges.  For security: The test of whether the user has sufficient authority within the CLIMSOFT software will always be whether they have admin access to the MySQL database they are connecting to.

We're now at the _"Progress with issue?"_ stage in our [governance process](https://github.com/opencdms/Climsoft/wiki/Open-Source-Contributions), we need agreement from the @opencdms/leaddevelopers

**Comment by isedwards** [comment]
@smachua @mhabimana - could you both vote on whether to proceed (or comment instead if there is anything you disagree with and want to discuss more)?

    +1 agree, -1 disagree, 0 abstain


My vote: 👍 +1 for me

**Comment by mhabimana** [comment]
@isedwards - I have a chat with @smachua on the above this afternoon and will be back to you.

**Comment by mhabimana** [comment]
- @smachua - We have agreed with @smachua that number 1 above should be done first before number 2 in this regards @smachua is working to fix the conflicts and then he will then push a new PR to GitHub for merging with the master;
- Re - whether to use MySQL or MariaDB we agreed that we should leave this optional to users since both were tested and are working very well( in short we leave the choice to users to use MySQL or MariaDB).
- It will be a good idea to create the setup using advance installer,then we will compare both Inno- setup and Advanced installer outputs and then decide on the way forward.

**Comment by isedwards** [comment]
Sounds good

**Comment by isedwards** [comment]
- [ ] Add a Welcome screen when Climsoft is launched for the first time
    - Welcome to Climsoft. Would you like to create a new database of connect to an existing database
- [ ] Add capability to create new database
- [ ] Ensure that default database for future logins is set to the same details as used during welcome
- [ ] When Climsoft is upgraded, the welcome screen should retain knowledge of the previous connections
- [ ] If a user provides server connection details we should be able to list available *Climsoft* databases through a combination of `SHOW DATABASES` and ensuring that the databases shown have the Climsoft schema (we shouldn't show non-Climsoft databases)
- [ ] Remove the instructions text file from directory
- [ ] Remove references to instructions text file from manual and installer
- [ ] Update manual to remove steps that are no longer required
- [ ] If the user requires more instructions, consider showing an image that illustrates server client architecture and illustrates that the database should be installed on the server and Climsoft desktop on the client.
- [ ] Start revering to Climsoft as "Climsoft Desktop"
